File: tc-v2-verify-go/ton_proof.go
package tcv2verify

import (
	"crypto/sha256"
	"encoding/binary"
)

// Contract B — TC_V2_TONPROOF_VERIFY_V1 (owner authentication, ton-proof-item-v2).
// Little-endian length/timestamp, no type discriminator, nested sha256. Owns its own
// encoders; shares no serialization helper with sign_data.go. encodeDomainLenB /
// encodeTimestampB are LITTLE-endian here vs big-endian in Contract A — the divergence
// that makes unifying the two a correctness bug.

const (
	tonProofPrefix = "ton-proof-item-v2/"
	tonProofOuter  = "ton-connect"
)

// TonProofInput is the input to Contract B.
type TonProofInput struct {
	Workchain    int32
	AddressHash  [32]byte
	Domain       string
	Timestamp    uint64
	ProofPayload string // the dApp nonce, signed as its literal string bytes (NOT decoded)
}

// TonProofDigest builds the Contract B message and returns its nested sha256 digest.
func TonProofDigest(in TonProofInput) [32]byte {
	domain := []byte(in.Domain)

	inner := make([]byte, 0, len(tonProofPrefix)+4+32+4+len(domain)+8+len(in.ProofPayload))
	inner = append(inner, tonProofPrefix...)
	inner = append(inner, encodeWorkchainB(in.Workchain)...)
	inner = append(inner, in.AddressHash[:]...)
	inner = append(inner, encodeDomainLenB(uint32(len(domain)))...)
	inner = append(inner, domain...)
	inner = append(inner, encodeTimestampB(in.Timestamp)...)
	inner = append(inner, in.ProofPayload...)
	innerHash := sha256.Sum256(inner)

	outer := make([]byte, 0, 2+len(tonProofOuter)+32)
	outer = append(outer, 0xff, 0xff)
	outer = append(outer, tonProofOuter...)
	outer = append(outer, innerHash[:]...)
	return sha256.Sum256(outer)
}

// Contract B field encoders — workchain big-endian; domain_len/timestamp LITTLE-endian.
// Deliberately NOT shared with sign_data.go.
func encodeWorkchainB(wc int32) []byte {
	b := make([]byte, 4)
	binary.BigEndian.PutUint32(b, uint32(wc))
	return b
}
func encodeDomainLenB(n uint32) []byte {
	b := make([]byte, 4)
	binary.LittleEndian.PutUint32(b, n)
	return b
}
func encodeTimestampB(ts uint64) []byte {
	b := make([]byte, 8)
	binary.LittleEndian.PutUint64(b, ts)
	return b
}

File: tc-v2-verify-rs/src/ton_proof.rs
//! Contract B — `TC_V2_TONPROOF_VERIFY_V1` (owner authentication, ton-proof-item-v2).
//!
//! LITTLE-endian length/timestamp, NO type discriminator, NESTED sha256. This module
//! owns its own field encoders. It shares NO serialization/endian/hash-pipeline helper
//! with `sign_data.rs` — see docs/spec/tc-v2-contract-boundaries.md. Note that
//! `encode_domain_length` / `encode_timestamp` here are little-endian, while Contract A's
//! are big-endian: this difference is exactly why the two must never be unified.

use crate::sha256::sha256;

const PROOF_PREFIX: &[u8] = b"ton-proof-item-v2/";
const OUTER_PREFIX: &[u8] = b"ton-connect";

pub struct TonProofInput<'a> {
    pub workchain: i32,
    pub address_hash: [u8; 32],
    pub domain: &'a str,
    pub timestamp: u64,
    /// The dApp nonce, signed as its LITERAL string bytes (NOT base64-decoded).
    pub proof_payload: &'a str,
}

/// Build the Contract B message and return its (nested) sha256 digest.
pub fn ton_proof_digest(input: &TonProofInput) -> [u8; 32] {
    let domain = input.domain.as_bytes();

    let mut inner = Vec::new();
    inner.extend_from_slice(PROOF_PREFIX);
    inner.extend_from_slice(&encode_workchain(input.workchain));
    inner.extend_from_slice(&input.address_hash);
    inner.extend_from_slice(&encode_domain_length(domain.len() as u32));
    inner.extend_from_slice(domain);
    inner.extend_from_slice(&encode_timestamp(input.timestamp));
    inner.extend_from_slice(input.proof_payload.as_bytes());

    let inner_hash = sha256(&inner);

    let mut outer = Vec::new();
    outer.extend_from_slice(&[0xff, 0xff]);
    outer.extend_from_slice(OUTER_PREFIX);
    outer.extend_from_slice(&inner_hash);
    sha256(&outer)
}

// --- Contract B field encoders. workchain BE; domain_len/timestamp LITTLE-endian.
//     NOT shared with Contract A. ---
fn encode_workchain(wc: i32) -> [u8; 4] {
    wc.to_be_bytes()
}
fn encode_domain_length(n: u32) -> [u8; 4] {
    n.to_le_bytes()
}
fn encode_timestamp(ts: u64) -> [u8; 8] {
    ts.to_le_bytes()
}
